### PR TITLE
Updated the readme to contain instructions to fake out the filesystem.

### PR DESCRIPTION
**Pull request by Skarlso**

### CODE REVIEW

File: README.markdown [hunk 1]
```diff
@@ -95,6 +95,65 @@ end
...
+TypeError:
+  superclass mismatch for class File
+```
+
+To prevent this, you'll need to require 'pp' *BEFORE* you require 'fakefs/spec_helpers' in your
+spec_helpers file. Like this:
+
+```ruby
+require 'pp'
+require 'fakefs/spec_helpers'
```
Comment (grosser) [inline README.markdown]: can you put the require into the spec_helper (or remove what causes the bug) and remove this note ?

File: README.markdown [hunk 2]
```diff
@@ -95,6 +95,65 @@ end
...
+
+describe "my spec" do
+  include FakeFS::SpecHelpers
+end
+```
+
+Mocking relative path in a Gem
```
Comment (grosser) [inline README.markdown]: can be simplified to `Mocking existing files` ?

File: README.markdown [hunk 3]
```diff
@@ -95,6 +95,65 @@ end
...
+------------------------------
+
+If you are working in a Gem and would like to load configuration files which sit with your gem in a
```
Comment (grosser) [inline README.markdown]: can be simplified to "If you want to load existing files" ?

File: README.markdown [hunk 4]
```diff
@@ -95,6 +95,65 @@ end
...
+cfg folder, you will need to clone your filesystem first. To do this, add the following line to your
+spec_helpers file.
+
+```ruby
+...
+shared_examples 'with faked out filesystem' do
+  before :each do
+    include FakeFS::SpecHelpers
```
Comment (grosser) [inline README.markdown]: should not include in an each ?

File: README.markdown [hunk 5]
```diff
@@ -95,6 +95,65 @@ end
...
+    FakeFS.activate!
+    FakeFS::FileSystem.clone(__dir__)
```
Comment (grosser) [inline README.markdown]: maybe `folder_to_clone` to be more generic / obvious where to edit ?

File: README.markdown [hunk 6]
```diff
@@ -95,6 +95,65 @@ end
...
+  end
+end
+...
+```
+
+And than if you have a test which needs to use a faked filesystem, like a configuration test, you'll
```
Comment (grosser) [inline README.markdown]: ... don't think this needs any extra comment, it's pretty obvious that it is setup above and then getting used here ... I'd even get rid of the shared example stuff to make it simpler ...

File: README.markdown [hunk 7]
```diff
@@ -95,6 +95,65 @@ end
...
+have to use this setup in the test like this.
+
+```ruby
+...
+describe MyModule::MySuperAwesomeClass, fakefs: true do
```
Comment (grosser) [inline README.markdown]: does the `fakefs: true` do anything ... why did you add it ?

File: README.markdown [hunk 8]
```diff
@@ -95,6 +95,65 @@ end
...
+  include_context 'with faked out filesystem'
+  ...
+```
+
+Once this is done, your configuration class will now look in the faked out filesystem, so you'll have
+to create whatever configuration file you have in a setup, like this.
+
+```ruby
+FakeFS do
+  FileUtils.mkdir_p(File.join(__dir__, '../cfg'))
```
Comment (grosser) [inline README.markdown]: I don't understand why you clone a directory up top and then create the file here anyway ?

File: README.markdown (hunk 1 above)
Comment (tovodeverett) [inline README.markdown]: I don't know if having some history helps, but this has come up before.  In issue https://github.com/fakefs/fakefs/issues/215 I mentioned that I used the solution from https://github.com/fakefs/fakefs/issues/99.  For what it's worth, I've been doing the `require 'pp'` after the `require 'fakefs/spec_helpers'` - I think it's just important to ensure the `require 'pp'` happens before any calls that actually use `fakefs`.

Comment (Skarlso) [inline README.markdown]: @grosser I have no idea what causes it. :/ This is the spec_helper. Or which spec_helper are you referring to?

File: README.markdown (hunk 7 above)
Comment (Skarlso) [inline README.markdown]: I think I was trying multiple things, and than that just got left there. :/ I will delete it. :)

File: README.markdown (hunk 8 above)
Comment (Skarlso) [inline README.markdown]: Right. I will re-word this part. I was trying to say, that, if you would like to overwrite the existing one, because you are entering test data. Like, if you have a public key file there, but you don't want to go through the hassle of it, just enter asdf.

**Comment by Skarlso** [comment]
Updated. Github is just crap right now, not showing outdated comments. :(

**Comment by grosser** [comment]
reviews still stay even if they are outdated ... I guess to show what was
discussed or so ...

On Sun, Oct 2, 2016 at 11:52 AM, Gergely Brautigam <notifications@github.com

File: README.markdown [hunk 9]
```diff
@@ -95,6 +95,62 @@ end
...
+TypeError:
+  superclass mismatch for class File
+```
+
+To prevent this, you'll need to require 'pp' *BEFORE* you require 'fakefs/spec_helpers' in your
+spec_helpers file. Like this:
+
+```ruby
+require 'pp'
+require 'fakefs/spec_helpers'
```
Comment (grosser) [inline README.markdown]: How about this ... sweet and short:

```
### FakeFs vs `pp` --- `TypeError: superclass mismatch for class File`

`pp` and `fakefs` collide, require `pp` before loading `fakefs`.
```

File: README.markdown [hunk 10]
```diff
@@ -95,6 +95,62 @@ end
...
+
+describe "my spec" do
+  include FakeFS::SpecHelpers
+end
+```
+
+Mocking existing files
+----------------------
+
+If you want to load existing files like configuration files which sit with your gem in a
```
Comment (grosser) [inline README.markdown]: how about: `If you want to use existing files in your fake fs, clone them first.`

Comment (grosser) [inline README.markdown]: and as example:

```Ruby
it "can use existing files" do
  FakeFS.activate!
  FakeFS::FileSystem.clone('<folder-to-clone>')
  expact(File.read('<folder-to-clone>/foo.yml')).to include("CLONED!")
end
```

Comment (Skarlso) [inline README.markdown]: What. How would it contained CLONED!? I'm not getting this one. :) 

Also, actually, I found that if I don't create the file, but clone my folder, it still complains that the file is not there. :/ But that must be something on my side then. :)

Comment (grosser) [inline README.markdown]: .... the `CLONED!` was mean as example ... I guess `content-from-original` would be better ...

Comment (grosser) [inline README.markdown]: also clone might be not working ... you can try adding a test-case for that ... would not be surprised if it does not work :D

Comment (Skarlso) [inline README.markdown]: Baaaaaaaaaaaah. :D

Comment (Skarlso) [inline README.markdown]: @grosser It actually does not work. :D I've just created a test in my project, removed all the other crap and just run this test:

```
    it 'correctly setup variables which are loaded from the config file' do
      FakeFS.activate!
      FakeFS::FileSystem.clone(__dir__)
      expect(mineconfig.loglevel).to eq('ERROR')
      expect(mineconfig.upload_path).to eq('/drop')
    end
```

And it faild with loading the config yml file. :) It tries to load it from the cfg folder. But because I'm running it with rake from the root folder, it should have successfully cloned my project.

Comment (grosser) [inline README.markdown]: hmmm so there is a bug in fakefs or just the example ?
... the Readme update now is what worked ?

Comment (Skarlso) [inline README.markdown]: I'm not a 100% sure here. :/ Do I need to setup anything else in order for the clone to work?

Comment (Skarlso) [inline README.markdown]: Damn. Works with absolute path. :/ So, I'm guessing __dir__ might be wrong here.

Comment (Skarlso) [inline README.markdown]: BTw, it's pretty slow. :/

**Comment by grosser** [comment]
... make the example use File.expand_path('../config', __FILE__) ... and I
guess try to add a test for relative paths working in a separate PR if it
looks like it's broken ...

**Comment by Skarlso** [comment]
I've got it. I was doing it from the wrong folder. This is how it's supposed to work:

```
      FakeFS.activate!
      FakeFS::FileSystem.clone(File.join(File.dirname(__FILE__), '../cfg'))
```

So it's cloning the cfg folder relative to the file it's running from. Which will be the spec folder. I was doing __dir__ which is not good because that's execution path relative so it was cloning the whole project which I don't want. So the clone is working fine. I'll update the README and my specs as well. :)

**Comment by grosser** [comment]
copies all files from that folder :D

alternatively you could use Dir.mktmpdir or this instead:

begin
old = File.read 'xxx'
File.write 'xxx', 'test'
... test stuff ...
ensure
File.write 'xxx', old
end

**Comment by Skarlso** [comment]
How about this? Also, here is my project in which I'm using fakefs. [AWS_Minecraft](https://github.com/Skarlso/aws_minecraft). Look into the spec folder. 

I have to go, it's super late here. :/ I'm continuing on this tomorrow. :) See ya!

File: README.markdown [hunk 11]
```diff
@@ -95,6 +95,46 @@ end
...
+end
+```
+
+Once this is done, your configuration class will now look in the faked out filesystem. For testing
+purposes if you would like to overwrite the content of a cloned file, do the following.
+
+```ruby
+FakeFS do
+  FileUtils.mkdir_p(File.join(__dir__, '../cfg'))
+  File.open(File.join(__dir__, '../cfg/config.yml'), 'w') do |f|
```
Comment (grosser) [inline README.markdown]: use `File.write` ... it's simpler ?

File: README.markdown [hunk 12]
```diff
@@ -95,6 +95,46 @@ end
...
+`pp` and `fakefs` collide, `require 'pp'` then `require 'fakefs'`.
+
+Mocking existing files
+----------------------
+
+If you want to use existing files in your fake fs, clone them first.
+
+```ruby
+it "can use existing files" do
+  FakeFS.activate!
```
Comment (grosser) [inline README.markdown]: should this use the block style instead ? ... otherwise that is a global change ?

File: README.markdown [hunk 13]
```diff
@@ -95,6 +95,46 @@ end
...
+end
+```
+
+For example, if you are running this from a spec folder, and would like to clone the cfg folder
+which is one step outside, you would do it like this:
+
+```ruby
+it "can use existing files" do
+  FakeFS.activate!
+  FakeFS::FileSystem.clone(File.join(File.dirname(__FILE__), '../cfg'))
```
Comment (grosser) [inline README.markdown]: `File.expand_path('../../cfg', __FILE__)` ?

File: README.markdown [hunk 14]
```diff
@@ -95,6 +95,46 @@ end
...
+
+```ruby
+it "can use existing files" do
+  FakeFS.activate!
+  FakeFS::FileSystem.clone('<folder-to-clone>')
+  expect(File.read('<folder-to-clone>/foo.yml')).to include("original-content-of-foo")
+end
+```
+
+For example, if you are running this from a spec folder, and would like to clone the cfg folder
```
Comment (grosser) [inline README.markdown]: these 2 last paragraphs seem kind of redundant ... maybe make the first example be `.clone(File.expand_path('../../config', __FILE)) + File.read('../config/test.yml')` ... then add `File.write("../config/test.yml", "... and change it!")` ... and everything should be obvious :)

**Comment by Skarlso** [comment]
Oh, you merged it before I addressed your new comments. :D thanks in any case. :)
